### ISSUE TITLE
npm@3: changing version no., doing npm install requires de-dupe

### ISSUE THREAD

**Issue by lukeapage** [npm/npm#8768]
I had a dependency on gulp-tslint 3.0.1-beta and the following structure
```
node_modules/
   gulp-tslint
   tslint
```

I then changed the package.json so gulp-tslint was at 3.1.0-beta, then did npm-install. I then got

```
node_modules/
    gulp-tslint/
        node_modules/
            tslint
```

so I had to do `npm dedupe` to sort it out, which worked.

Is npm getting confused when sub modules versions are changing and therefore not flattening? or is this expected behaviour? nothing else I have depends on the sub module.

**Comment by iarna** [comment]
This is a side effect of the upgrade algorithm:

It starts by replacing the `gulp-tslint` entry in the tree. Then it walks its deps to make sure they're up to date– it finds that it depends on `tslint@~2.4.2` but that's not compatible with the one currently installed, so it gives `gulp-tslint` its own copy.

Then, much later, we walk the entire tree and look for orphaned modules and we find `tslint@2.2.0-beta` left over from the older `gulp-tslint`. It's no longer required by anything so its removed. 

Adding a hoisting step (much like what dedupe does) to the pruner would be the best course of action I think. Eg, if we've just removed a module, go look for other modules with the same name and pick one to promote up the tree.

**Comment by iarna** [comment]
This was fixed with https://github.com/npm/npm/commit/971fd47 (https://github.com/npm/npm/pull/9929), which landed last week in 3.3.10!
